--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -19,7 +_,11 @@
 
 paperCheckstyle {
     customJavadocTags = projectCustomJavadocTags
-    directoriesToSkipFile = layout.projectDirectory.file(".checkstyle/ignored_directories.txt")
+    directoriesToSkipFile = layout.projectDirectory.file("../paper-api/.checkstyle/ignored_directories.txt")
+}
+
+tasks.named<io.papermc.paperweight.checkstyle.tasks.MergeCheckstyleConfigs>("mergeCheckstyleConfigs") {
+    overrideConfigFile.set(layout.projectDirectory.file("../paper-api/.checkstyle/checkstyle.xml"))
 }
 
 val annotationsVersion = "26.0.2"
@@ -105,10 +_,10 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
 
     // checkstyle
-    checkstyle(project(":paper-checkstyle"))
+    checkstyle(project(":folia-checkstyle"))
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath()
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -118,6 +_,18 @@
     main {
         java {
             srcDir(generatedDir)
+            srcDir(file("../paper-api/src/main/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/main/resources"))
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
         }
     }
 }
@@ -200,7 +_,7 @@
 
 tasks.withType<Javadoc>().configureEach {
     val options = options as StandardJavadocDocletOptions
-    options.overview = "src/main/javadoc/overview.html"
+    options.overview = "../paper-api/src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -227,11 +_,11 @@
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../paper-api/src/main/javadoc").withPropertyName("javadoc-sourceset")
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../paper-api/src/main/javadoc") {
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
