From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 17 Feb 2026 20:12:19 +0900
Subject: [PATCH] Add custom heightmap types


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftHeightMap.java b/src/main/java/org/bukkit/craftbukkit/CraftHeightMap.java
index 84be88047c441679a0f27302935aa495e73883e0..9b5d545e2897ecf51273726053f406def3a6d6af 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftHeightMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftHeightMap.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit;
 
+import net.minecraft.world.level.levelgen.Heightmap;
 import org.bukkit.HeightMap;
 
 public final class CraftHeightMap {
@@ -7,19 +8,42 @@ public final class CraftHeightMap {
     private CraftHeightMap() {
     }
 
-    public static net.minecraft.world.level.levelgen.Heightmap.Types toNMS(HeightMap bukkitHeightMap) {
+    public static Heightmap.Types toNMS(HeightMap bukkitHeightMap) {
         return switch (bukkitHeightMap) {
             case MOTION_BLOCKING_NO_LEAVES ->
-                net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING_NO_LEAVES;
-            case OCEAN_FLOOR -> net.minecraft.world.level.levelgen.Heightmap.Types.OCEAN_FLOOR;
-            case OCEAN_FLOOR_WG -> net.minecraft.world.level.levelgen.Heightmap.Types.OCEAN_FLOOR_WG;
-            case WORLD_SURFACE -> net.minecraft.world.level.levelgen.Heightmap.Types.WORLD_SURFACE;
-            case WORLD_SURFACE_WG -> net.minecraft.world.level.levelgen.Heightmap.Types.WORLD_SURFACE_WG;
-            case MOTION_BLOCKING -> net.minecraft.world.level.levelgen.Heightmap.Types.MOTION_BLOCKING;
+                Heightmap.Types.MOTION_BLOCKING_NO_LEAVES;
+            case OCEAN_FLOOR -> Heightmap.Types.OCEAN_FLOOR;
+            case OCEAN_FLOOR_WG -> Heightmap.Types.OCEAN_FLOOR_WG;
+            case WORLD_SURFACE -> Heightmap.Types.WORLD_SURFACE;
+            case WORLD_SURFACE_WG -> Heightmap.Types.WORLD_SURFACE_WG;
+            case MOTION_BLOCKING -> Heightmap.Types.MOTION_BLOCKING;
+
+            // Aetheria start - Converter for custom heightmap types (API -> Minecraft)
+            case AETHERIA_OVERWORLD_MOTION_BLOCKING -> Heightmap.Types.AETHERIA_OVERWORLD_MOTION_BLOCKING;
+            case AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES -> Heightmap.Types.AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES;
+            case AETHERIA_OVERWORLD_OCEAN_FLOOR -> Heightmap.Types.AETHERIA_OVERWORLD_OCEAN_FLOOR;
+            case AETHERIA_OVERWORLD_OCEAN_FLOOR_WG -> Heightmap.Types.AETHERIA_OVERWORLD_OCEAN_FLOOR_WG;
+            case AETHERIA_OVERWORLD_WORLD_SURFACE -> Heightmap.Types.AETHERIA_OVERWORLD_WORLD_SURFACE;
+            case AETHERIA_OVERWORLD_WORLD_SURFACE_WG -> Heightmap.Types.AETHERIA_OVERWORLD_WORLD_SURFACE_WG;
+
+            case AETHERIA_NETHER_MOTION_BLOCKING -> Heightmap.Types.AETHERIA_NETHER_MOTION_BLOCKING;
+            case AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES -> Heightmap.Types.AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES;
+            case AETHERIA_NETHER_OCEAN_FLOOR -> Heightmap.Types.AETHERIA_NETHER_OCEAN_FLOOR;
+            case AETHERIA_NETHER_OCEAN_FLOOR_WG -> Heightmap.Types.AETHERIA_NETHER_OCEAN_FLOOR_WG;
+            case AETHERIA_NETHER_WORLD_SURFACE -> Heightmap.Types.AETHERIA_NETHER_WORLD_SURFACE;
+            case AETHERIA_NETHER_WORLD_SURFACE_WG -> Heightmap.Types.AETHERIA_NETHER_WORLD_SURFACE_WG;
+
+            case AETHERIA_END_MOTION_BLOCKING -> Heightmap.Types.AETHERIA_END_MOTION_BLOCKING;
+            case AETHERIA_END_MOTION_BLOCKING_NO_LEAVES -> Heightmap.Types.AETHERIA_END_MOTION_BLOCKING_NO_LEAVES;
+            case AETHERIA_END_OCEAN_FLOOR -> Heightmap.Types.AETHERIA_END_OCEAN_FLOOR;
+            case AETHERIA_END_OCEAN_FLOOR_WG -> Heightmap.Types.AETHERIA_END_OCEAN_FLOOR_WG;
+            case AETHERIA_END_WORLD_SURFACE -> Heightmap.Types.AETHERIA_END_WORLD_SURFACE;
+            case AETHERIA_END_WORLD_SURFACE_WG -> Heightmap.Types.AETHERIA_END_WORLD_SURFACE_WG;
+            // Aetheria end - Converter for custom heightmap types (API -> Minecraft)
         };
     }
 
-    public static HeightMap fromNMS(net.minecraft.world.level.levelgen.Heightmap.Types nmsHeightMapType) {
+    public static HeightMap fromNMS(Heightmap.Types nmsHeightMapType) {
         return switch (nmsHeightMapType) {
             case WORLD_SURFACE_WG -> HeightMap.WORLD_SURFACE_WG;
             case WORLD_SURFACE -> HeightMap.WORLD_SURFACE;
@@ -27,6 +51,29 @@ public final class CraftHeightMap {
             case OCEAN_FLOOR -> HeightMap.OCEAN_FLOOR;
             case MOTION_BLOCKING_NO_LEAVES -> HeightMap.MOTION_BLOCKING_NO_LEAVES;
             case MOTION_BLOCKING -> HeightMap.MOTION_BLOCKING;
+
+            // Aetheria start - Converter for custom heightmap types (Minecraft -> API)
+            case AETHERIA_OVERWORLD_WORLD_SURFACE_WG -> HeightMap.AETHERIA_OVERWORLD_WORLD_SURFACE_WG;
+            case AETHERIA_OVERWORLD_WORLD_SURFACE -> HeightMap.AETHERIA_OVERWORLD_WORLD_SURFACE;
+            case AETHERIA_OVERWORLD_OCEAN_FLOOR_WG ->  HeightMap.AETHERIA_OVERWORLD_OCEAN_FLOOR_WG;
+            case AETHERIA_OVERWORLD_OCEAN_FLOOR ->  HeightMap.AETHERIA_OVERWORLD_OCEAN_FLOOR;
+            case AETHERIA_OVERWORLD_MOTION_BLOCKING -> HeightMap.AETHERIA_OVERWORLD_MOTION_BLOCKING;
+            case AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES -> HeightMap.AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES;
+
+            case AETHERIA_NETHER_WORLD_SURFACE_WG -> HeightMap.AETHERIA_NETHER_WORLD_SURFACE_WG;
+            case AETHERIA_NETHER_WORLD_SURFACE -> HeightMap.AETHERIA_NETHER_WORLD_SURFACE;
+            case AETHERIA_NETHER_OCEAN_FLOOR_WG -> HeightMap.AETHERIA_NETHER_OCEAN_FLOOR_WG;
+            case AETHERIA_NETHER_OCEAN_FLOOR -> HeightMap.AETHERIA_NETHER_OCEAN_FLOOR;
+            case AETHERIA_NETHER_MOTION_BLOCKING -> HeightMap.AETHERIA_NETHER_MOTION_BLOCKING;
+            case AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES -> HeightMap.AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES;
+
+            case AETHERIA_END_WORLD_SURFACE_WG -> HeightMap.AETHERIA_END_WORLD_SURFACE_WG;
+            case AETHERIA_END_WORLD_SURFACE -> HeightMap.AETHERIA_END_WORLD_SURFACE;
+            case AETHERIA_END_OCEAN_FLOOR_WG -> HeightMap.AETHERIA_END_OCEAN_FLOOR_WG;
+            case AETHERIA_END_OCEAN_FLOOR -> HeightMap.AETHERIA_END_OCEAN_FLOOR;
+            case AETHERIA_END_MOTION_BLOCKING -> HeightMap.AETHERIA_END_MOTION_BLOCKING;
+            case AETHERIA_END_MOTION_BLOCKING_NO_LEAVES -> HeightMap.AETHERIA_END_MOTION_BLOCKING_NO_LEAVES;
+            // Aetheria end - Converter for custom heightmap types (Minecraft -> API)
         };
     }
 }
