From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 3 Mar 2026 17:50:11 +0900
Subject: [PATCH] Server item registry


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 121babcfdda53e8cade0edfc7f8bd4af56db9abe..a2993d01c9c83d8ff332312f3534c24d566c7b5d 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -30,6 +30,9 @@ import java.util.IdentityHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import net.azisaba.vanilife.item.ServerItem;
+import net.azisaba.vanilife.server.item.ServerItemImpl;
+import net.azisaba.vanilife.server.registry.data.ServerItemRegistryEntryImpl;
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.resources.ResourceKey;
@@ -147,8 +150,11 @@ public final class PaperRegistries {
             start(Registries.ENTITY_TYPE, RegistryKey.ENTITY_TYPE).apiOnly(PaperSimpleRegistry::entityType),
             start(Registries.PARTICLE_TYPE, RegistryKey.PARTICLE_TYPE).apiOnly(PaperSimpleRegistry::particleType),
             start(Registries.POTION, RegistryKey.POTION).apiOnly(PaperSimpleRegistry::potion),
-            start(Registries.MEMORY_MODULE_TYPE, RegistryKey.MEMORY_MODULE_TYPE).apiOnly(() -> org.bukkit.Registry.MEMORY_MODULE_TYPE)
+            start(Registries.MEMORY_MODULE_TYPE, RegistryKey.MEMORY_MODULE_TYPE).apiOnly(() -> org.bukkit.Registry.MEMORY_MODULE_TYPE),
             // End generate - RegistryDefinitions
+
+            // Vanilife Registries
+            start(Registries.SERVER_ITEM, RegistryKey.SERVER_ITEM).craft(ServerItem.class, ServerItemImpl::new).writable(ServerItemRegistryEntryImpl.BuilderImpl::new)
         );
         final Map<RegistryKey<?>, RegistryEntry<?, ?>> byRegistryKey = new IdentityHashMap<>(REGISTRY_ENTRIES.size());
         final Map<ResourceKey<?>, RegistryEntry<?, ?>> byResourceKey = new IdentityHashMap<>(REGISTRY_ENTRIES.size());
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 002682049fe890333794072e1e2717004f2ad0cf..a13e672606c16ccb136adac06ef48d4a5aa0ac4e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -1,13 +1,19 @@
 package org.bukkit.craftbukkit.inventory;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.ImmutableMap;
 import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.registry.RegistryAccess;
+import io.papermc.paper.registry.RegistryKey;
+import io.papermc.paper.registry.TypedKey;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.item.ServerItem;
 import net.kyori.adventure.text.Component;
 import net.minecraft.advancements.criterion.DataComponentMatchers;
 import net.minecraft.advancements.criterion.ItemPredicate;
@@ -26,6 +32,7 @@ import net.minecraft.world.item.component.CustomData;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.enchantment.ItemEnchantments;
 import org.bukkit.Material;
+import org.bukkit.Registry;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.craftbukkit.enchantments.CraftEnchantment;
 import org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer;
@@ -35,7 +42,9 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.material.MaterialData;
 import org.bukkit.persistence.PersistentDataContainer;
+import org.bukkit.persistence.PersistentDataType;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 @DelegateDeserialization(ItemStack.class)
 public final class CraftItemStack extends ItemStack {
@@ -482,6 +491,38 @@ public final class CraftItemStack extends ItemStack {
     }
     // Paper end
 
+    @Modified(Op.ADD_METHOD)
+    @Override
+    public @Nullable ServerItem serverItem() {
+        final TypedKey<ServerItem> typedKey = this.serverItemKey();
+        if (typedKey == null) {
+            return null;
+        }
+
+        final Registry<ServerItem> registry = RegistryAccess.registryAccess().getRegistry(RegistryKey.SERVER_ITEM);
+        return registry.get(typedKey);
+    }
+
+    @Modified(Op.ADD_METHOD)
+    @Override
+    public @Nullable TypedKey<ServerItem> serverItemKey() {
+        return Optional.ofNullable(getPersistentDataContainer().get(ServerItem.TYPE_KEY, PersistentDataType.STRING))
+                .map(RegistryKey.SERVER_ITEM::typedKey)
+                .orElse(null);
+    }
+
+    @Modified(Op.ADD_METHOD)
+    @Override
+    public boolean isOf(final @NotNull ServerItem type) {
+        return this.isOf(RegistryKey.SERVER_ITEM.typedKey(type.key()));
+    }
+
+    @Modified(Op.ADD_METHOD)
+    @Override
+    public boolean isOf(final @NotNull TypedKey<ServerItem> type) {
+        return Objects.equals(this.serverItemKey(), type);
+    }
+
     public static final String PDC_CUSTOM_DATA_KEY = "PublicBukkitValues";
     private net.minecraft.nbt.CompoundTag getPdcTag() {
         if (this.handle == null) {
