From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 3 Mar 2026 17:50:11 +0900
Subject: [PATCH] Server item registry


diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistries.java b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
index 121babcfdda53e8cade0edfc7f8bd4af56db9abe..bcae618f4d17578a9918dceb81b1cf92be6ea916 100644
--- a/src/main/java/io/papermc/paper/registry/PaperRegistries.java
+++ b/src/main/java/io/papermc/paper/registry/PaperRegistries.java
@@ -30,6 +30,9 @@ import java.util.IdentityHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import net.azisaba.vanilife.item.ServerItemType;
+import net.azisaba.vanilife.server.item.VanilifeServerItemType;
+import net.azisaba.vanilife.server.registry.data.VanilifeServerItemTypeRegistryEntry;
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.resources.ResourceKey;
@@ -147,8 +150,11 @@ public final class PaperRegistries {
             start(Registries.ENTITY_TYPE, RegistryKey.ENTITY_TYPE).apiOnly(PaperSimpleRegistry::entityType),
             start(Registries.PARTICLE_TYPE, RegistryKey.PARTICLE_TYPE).apiOnly(PaperSimpleRegistry::particleType),
             start(Registries.POTION, RegistryKey.POTION).apiOnly(PaperSimpleRegistry::potion),
-            start(Registries.MEMORY_MODULE_TYPE, RegistryKey.MEMORY_MODULE_TYPE).apiOnly(() -> org.bukkit.Registry.MEMORY_MODULE_TYPE)
+            start(Registries.MEMORY_MODULE_TYPE, RegistryKey.MEMORY_MODULE_TYPE).apiOnly(() -> org.bukkit.Registry.MEMORY_MODULE_TYPE),
             // End generate - RegistryDefinitions
+
+            // Vanilife Registries
+            start(Registries.SERVER_ITEM, RegistryKey.SERVER_ITEM).craft(ServerItemType.class, VanilifeServerItemType::new).writable(VanilifeServerItemTypeRegistryEntry.VanilifeBuilder::new)
         );
         final Map<RegistryKey<?>, RegistryEntry<?, ?>> byRegistryKey = new IdentityHashMap<>(REGISTRY_ENTRIES.size());
         final Map<ResourceKey<?>, RegistryEntry<?, ?>> byResourceKey = new IdentityHashMap<>(REGISTRY_ENTRIES.size());
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 002682049fe890333794072e1e2717004f2ad0cf..3d3b3d4ff124faf90e7e19fcf0590368c91aadd2 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -1,13 +1,17 @@
 package org.bukkit.craftbukkit.inventory;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.ImmutableMap;
 import io.papermc.paper.adventure.PaperAdventure;
 import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.item.ServerItemType;
+import net.azisaba.vanilife.server.item.VanilifeServerItemType;
+import net.kyori.adventure.key.Key;
 import net.kyori.adventure.text.Component;
 import net.minecraft.advancements.criterion.DataComponentMatchers;
 import net.minecraft.advancements.criterion.ItemPredicate;
@@ -35,7 +39,9 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.material.MaterialData;
 import org.bukkit.persistence.PersistentDataContainer;
+import org.bukkit.persistence.PersistentDataType;
 import org.jetbrains.annotations.NotNull;
+import org.jspecify.annotations.NonNull;
 
 @DelegateDeserialization(ItemStack.class)
 public final class CraftItemStack extends ItemStack {
@@ -482,6 +488,15 @@ public final class CraftItemStack extends ItemStack {
     }
     // Paper end
 
+    @Modified(Op.ADD_METHOD)
+    @Override
+    public boolean isOf(final @NonNull ServerItemType type) {
+        final Key key = Optional.ofNullable(getPersistentDataContainer().get(VanilifeServerItemType.TYPE_KEY, PersistentDataType.STRING))
+                .map(Key::key)
+                .orElse(null);
+        return key != null && type.key().equals(key);
+    }
+
     public static final String PDC_CUSTOM_DATA_KEY = "PublicBukkitValues";
     private net.minecraft.nbt.CompoundTag getPdcTag() {
         if (this.handle == null) {
