From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 17 Feb 2026 20:14:12 +0900
Subject: [PATCH] Add custom heightmap types


diff --git a/src/main/java/org/bukkit/HeightMap.java b/src/main/java/org/bukkit/HeightMap.java
index 344b2b5d9207d2645bc5417d1ec00dd0a0b95604..f334296d103415b8435a3e046e665d11a723a92d 100644
--- a/src/main/java/org/bukkit/HeightMap.java
+++ b/src/main/java/org/bukkit/HeightMap.java
@@ -31,4 +31,27 @@ public enum HeightMap {
      * The highest non-air block, for worldgen.
      */
     WORLD_SURFACE_WG,
+
+    // Aetheria start - Custom heightmap types by Aetheria layer
+    AETHERIA_OVERWORLD_MOTION_BLOCKING,
+    AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES,
+    AETHERIA_OVERWORLD_OCEAN_FLOOR,
+    AETHERIA_OVERWORLD_OCEAN_FLOOR_WG,
+    AETHERIA_OVERWORLD_WORLD_SURFACE,
+    AETHERIA_OVERWORLD_WORLD_SURFACE_WG,
+
+    AETHERIA_NETHER_MOTION_BLOCKING,
+    AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES,
+    AETHERIA_NETHER_OCEAN_FLOOR,
+    AETHERIA_NETHER_OCEAN_FLOOR_WG,
+    AETHERIA_NETHER_WORLD_SURFACE,
+    AETHERIA_NETHER_WORLD_SURFACE_WG,
+
+    AETHERIA_END_MOTION_BLOCKING,
+    AETHERIA_END_MOTION_BLOCKING_NO_LEAVES,
+    AETHERIA_END_OCEAN_FLOOR,
+    AETHERIA_END_OCEAN_FLOOR_WG,
+    AETHERIA_END_WORLD_SURFACE,
+    AETHERIA_END_WORLD_SURFACE_WG,
+    // Aetheria end - Custom heightmap types by Aetheria layer
 }
