From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 3 Mar 2026 17:50:15 +0900
Subject: [PATCH] Server item registry


diff --git a/src/main/java/io/papermc/paper/registry/RegistryKey.java b/src/main/java/io/papermc/paper/registry/RegistryKey.java
index dcee3d88f5b77cdd5c7cbc07f73f54cc548d2f94..d877bda2d5a556a4ce09c9c8bfb85def2b9d8f47 100644
--- a/src/main/java/io/papermc/paper/registry/RegistryKey.java
+++ b/src/main/java/io/papermc/paper/registry/RegistryKey.java
@@ -3,6 +3,10 @@ package io.papermc.paper.registry;
 import io.papermc.paper.datacomponent.DataComponentType;
 import io.papermc.paper.dialog.Dialog;
 import io.papermc.paper.registry.tag.TagKey;
+import net.azisaba.vanilife.Vanilife;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.item.ServerItemType;
 import net.kyori.adventure.key.Key;
 import net.kyori.adventure.key.KeyPattern;
 import net.kyori.adventure.key.Keyed;
@@ -241,6 +245,9 @@ public sealed interface RegistryKey<T> extends Keyed permits RegistryKeyImpl {
     RegistryKey<PotionType> POTION = create("potion");
     RegistryKey<MemoryKey<?>> MEMORY_MODULE_TYPE = create("memory_module_type");
 
+    @Modified(Op.ADD_FIELD)
+    RegistryKey<ServerItemType> SERVER_ITEM = create(Key.key(Vanilife.NAMESPACE, "server_item"));
+
     /**
      * Constructs a new {@link TypedKey} for this registry given the typed key's key.
      *
diff --git a/src/main/java/io/papermc/paper/registry/RegistryKeyImpl.java b/src/main/java/io/papermc/paper/registry/RegistryKeyImpl.java
index ac68dfce754dc7e014bb31bba32d9b246ffd411c..084a31cec9720c409b3e814723509ea952c5a620 100644
--- a/src/main/java/io/papermc/paper/registry/RegistryKeyImpl.java
+++ b/src/main/java/io/papermc/paper/registry/RegistryKeyImpl.java
@@ -2,6 +2,8 @@ package io.papermc.paper.registry;
 
 import com.google.common.collect.Sets;
 import java.util.Set;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
 import net.kyori.adventure.key.Key;
 import org.intellij.lang.annotations.Subst;
 import org.jspecify.annotations.NullMarked;
@@ -29,6 +31,13 @@ record RegistryKeyImpl<T>(Key key) implements RegistryKey<T> {
         return registryKey;
     }
 
+    @Modified(Op.ADD_METHOD)
+    static <T> RegistryKey<T> create(final Key key) {
+        final RegistryKey<T> registryKey = new RegistryKeyImpl<>(key);
+        REGISTRY_KEYS.add(registryKey);
+        return registryKey;
+    }
+
     // creates the key without adding to the internal set of keys
     static <T> RegistryKey<T> createInternal(@Subst("some_key") final String key) {
         return new RegistryKeyImpl<>(Key.key(Key.MINECRAFT_NAMESPACE, key));
diff --git a/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java b/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
index 6ae47dac3496e2abf78b13ba1ae5dd0f3b2dd158..719998951b5fb0aeb3d195efdcdfa902ad94d951 100644
--- a/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
+++ b/src/main/java/io/papermc/paper/registry/event/RegistryEvents.java
@@ -17,6 +17,10 @@ import io.papermc.paper.registry.data.PigVariantRegistryEntry;
 import io.papermc.paper.registry.data.WolfVariantRegistryEntry;
 import io.papermc.paper.registry.data.ZombieNautilusVariantRegistryEntry;
 import io.papermc.paper.registry.data.dialog.DialogRegistryEntry;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.item.ServerItemType;
+import net.azisaba.vanilife.registry.data.ServerItemTypeRegistryEntry;
 import org.bukkit.Art;
 import org.bukkit.GameEvent;
 import org.bukkit.JukeboxSong;
@@ -57,6 +61,8 @@ public final class RegistryEvents {
     public static final RegistryEventProvider<ZombieNautilus.Variant, ZombieNautilusVariantRegistryEntry.Builder> ZOMBIE_NAUTILUS_VARIANT = create(RegistryKey.ZOMBIE_NAUTILUS_VARIANT);
     public static final RegistryEventProvider<Dialog, DialogRegistryEntry.Builder> DIALOG = create(RegistryKey.DIALOG);
     // End generate - RegistryEvents
+    @Modified(Op.ADD_FIELD)
+    public static final RegistryEventProvider<ServerItemType, ServerItemTypeRegistryEntry.Builder> SERVER_ITEM = create(RegistryKey.SERVER_ITEM);
 
     private RegistryEvents() {
     }
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index f20f072b0978b3ef4b3f36c952d469e0d8d5cd22..398ee67be1e872f94c86df9a8d202fa6883c73c2 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -3,13 +3,14 @@ package org.bukkit.inventory;
 import com.google.common.base.Preconditions;
 import io.papermc.paper.datacomponent.DataComponentHolder;
 import io.papermc.paper.registry.RegistryKey;
-import java.util.LinkedHashMap;
 import java.util.Locale;
 import java.util.Map;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.item.ServerItemType;
 import net.kyori.adventure.text.Component;
-import net.kyori.adventure.text.event.HoverEvent;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.NamespacedKey;
@@ -66,6 +67,16 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
     }
     // Paper end
 
+    @Modified(Op.ADD_METHOD)
+    public static @NotNull ItemStack of(final @NotNull ServerItemType type) {
+        return of(type, 1);
+    }
+
+    @Modified(Op.ADD_METHOD)
+    public static @NotNull ItemStack of(final @NotNull ServerItemType type, final int amount) {
+        return type.createItemStack(amount);
+    }
+
     // Paper start - pdc
     /**
      * @see #editPersistentDataContainer(Consumer)
