--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -679,7 +_,7 @@
     }
 
     private static void dropOrPlaceInInventory(Player player, ItemStack stack) {
-        boolean flag = player.isRemoved() && player.getRemovalReason() != Entity.RemovalReason.CHANGED_DIMENSION;
+        boolean flag = !player.isAlive(); // player.isRemoved() && player.getRemovalReason() != Entity.RemovalReason.CHANGED_DIMENSION; // Folia - region threading
         boolean flag1 = player instanceof ServerPlayer serverPlayer && serverPlayer.hasDisconnected();
         if (flag || flag1) {
             player.drop(stack, false);
