From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rothes <449181985@qq.com>
Date: Thu, 19 Mar 2026 10:12:55 +0800
Subject: [PATCH] Fix vanish api sync error


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index d94c0d15f62b67378669d4ce7252d99b8f743fa5..1574ead80189fbd4e4f6d677992f3ea5671bf844 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -307,6 +307,7 @@ public final class RegionizedPlayerChunkLoader {
         TickThread.ensureTickThread("Cannot tick player chunk loader async");
         long currTime = System.nanoTime();
         for (final ServerPlayer player : new java.util.ArrayList<>(this.world.getLocalPlayers())) { // Folia - region threding
+            player.getBukkitEntity().onTick(); // Folia - Fix vanish api sync error
             final PlayerChunkLoaderData loader = ((ChunkSystemServerPlayer)player).moonrise$getChunkLoader();
             if (loader == null || loader.removed || loader.world != this.world) {
                 // not our problem anymore
