From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tksimeji <tksimeji@outlook.com>
Date: Tue, 3 Mar 2026 17:49:56 +0900
Subject: [PATCH] Server item registry


diff --git a/io/papermc/paper/threadedregions/ThreadedRegionizer.java b/io/papermc/paper/threadedregions/ThreadedRegionizer.java
index 604385af903845d966382ad0a4168798e4ed4a0e..b7c4500a40035ead2026e8f2fd4259926137ab7e 100644
--- a/io/papermc/paper/threadedregions/ThreadedRegionizer.java
+++ b/io/papermc/paper/threadedregions/ThreadedRegionizer.java
@@ -617,7 +617,7 @@ public final class ThreadedRegionizer<R extends ThreadedRegionizer.ThreadedRegio
             newRegions.add(currRegion);
         }
 
-        // now we have split the regions into separate parts, we can split recalculate
+        // now we have split the regions into separate conditionedParts, we can split recalculate
 
         if (newRegions.size() == 1) {
             // no need to split anything, we're done here
diff --git a/net/minecraft/core/registries/BuiltInRegistries.java b/net/minecraft/core/registries/BuiltInRegistries.java
index d3822f1713f588e9c7a41ecb408db99689f3d2ba..00d89f0cb74392227fb6e49251f8b71c368c0b9d 100644
--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -8,6 +8,9 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.registry.data.ServerItemRegistryEntry;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.advancements.CriterionTrigger;
 import net.minecraft.advancements.criterion.EntitySubPredicate;
@@ -355,6 +358,8 @@ public class BuiltInRegistries {
     public static final Registry<AttributeType<?>> ATTRIBUTE_TYPE = registerSimple(Registries.ATTRIBUTE_TYPE, AttributeTypes::bootstrap);
     public static final Registry<MapCodec<? extends SlotSource>> SLOT_SOURCE_TYPE = registerSimple(Registries.SLOT_SOURCE_TYPE, SlotSources::bootstrap);
     public static final Registry<Consumer<GameTestHelper>> TEST_FUNCTION = registerSimple(Registries.TEST_FUNCTION, BuiltinTestFunctions::bootstrap);
+    @Modified(Op.ADD_FIELD)
+    public static final Registry<ServerItemRegistryEntry> SERVER_ITEM = registerSimple(Registries.SERVER_ITEM, (registry) -> new Object());
     public static final Registry<? extends Registry<?>> REGISTRY = WRITABLE_REGISTRY;
     // Paper start - add built-in registry conversions
     public static final io.papermc.paper.registry.data.util.Conversions STATIC_ACCESS_CONVERSIONS = new io.papermc.paper.registry.data.util.Conversions(
diff --git a/net/minecraft/core/registries/Registries.java b/net/minecraft/core/registries/Registries.java
index ab7f29272be21d2815ccb5add96def059442f250..7b443c0b0b8cef92ab001ebdc785cdaf6ca25c2c 100644
--- a/net/minecraft/core/registries/Registries.java
+++ b/net/minecraft/core/registries/Registries.java
@@ -2,6 +2,10 @@ package net.minecraft.core.registries;
 
 import com.mojang.serialization.MapCodec;
 import java.util.function.Consumer;
+import net.azisaba.vanilife.Vanilife;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.registry.data.ServerItemRegistryEntry;
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.CriterionTrigger;
 import net.minecraft.advancements.criterion.EntitySubPredicate;
@@ -295,6 +299,8 @@ public class Registries {
     public static final ResourceKey<Registry<LootItemCondition>> PREDICATE = createRegistryKey("predicate");
     public static final ResourceKey<Registry<Advancement>> ADVANCEMENT = createRegistryKey("advancement");
     public static final ResourceKey<Registry<Recipe<?>>> RECIPE = createRegistryKey("recipe");
+    @Modified(Op.ADD_FIELD)
+    public static final ResourceKey<Registry<ServerItemRegistryEntry>> SERVER_ITEM = ResourceKey.createRegistryKey(Identifier.fromNamespaceAndPath(Vanilife.NAMESPACE, "server_item"));
 
     public static ResourceKey<Level> levelStemToLevel(ResourceKey<LevelStem> levelStem) {
         return ResourceKey.create(DIMENSION, levelStem.identifier());
diff --git a/net/minecraft/network/chat/contents/TranslatableContents.java b/net/minecraft/network/chat/contents/TranslatableContents.java
index 8da489e8179f772ed439360c3451fdff15f0dc22..bbc2de522b2b6702a7b2b0f85a9ff767cdf26e46 100644
--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -178,7 +178,7 @@ public class TranslatableContents implements ComponentContents {
 
     @Override
     public <T> Optional<T> visit(FormattedText.ContentConsumer<T> contentConsumer) {
-        // Paper start - Count visited parts
+        // Paper start - Count visited conditionedParts
         try {
             return this.visit(new TranslatableContentConsumer<>(contentConsumer));
         } catch (IllegalArgumentException ignored) {
@@ -187,7 +187,7 @@ public class TranslatableContents implements ComponentContents {
     }
 
     private <T> Optional<T> visit(TranslatableContentConsumer<T> contentConsumer) {
-        // Paper end - Count visited parts
+        // Paper end - Count visited conditionedParts
         this.decompose();
 
         for (FormattedText formattedText : this.decomposedParts) {
@@ -200,7 +200,7 @@ public class TranslatableContents implements ComponentContents {
         return Optional.empty();
     }
 
-    // Paper start - Count visited parts
+    // Paper start - Count visited conditionedParts
     private static final class TranslatableContentConsumer<T> implements FormattedText.ContentConsumer<T> {
         private static final IllegalArgumentException NESTED_TOO_LONG = new IllegalArgumentException("Too long");
 
@@ -219,7 +219,7 @@ public class TranslatableContents implements ComponentContents {
             return this.visitor.accept(asString);
         }
     }
-    // Paper end - Count visited parts
+    // Paper end - Count visited conditionedParts
 
     @Override
     public MutableComponent resolve(@Nullable CommandSourceStack source, @Nullable Entity entity, int recursionDepth) throws CommandSyntaxException {
diff --git a/net/minecraft/resources/RegistryDataLoader.java b/net/minecraft/resources/RegistryDataLoader.java
index 4a700aa0e68044bb950710be679861230ab5223d..70ea51f01b9f264fe8d76f512555e4e7879af58e 100644
--- a/net/minecraft/resources/RegistryDataLoader.java
+++ b/net/minecraft/resources/RegistryDataLoader.java
@@ -21,10 +21,11 @@ import java.util.Map.Entry;
 import java.util.function.BiConsumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
-
-import net.azisaba.vanilife.VanilifeBiomes;
-import net.azisaba.vanilife.VanilifeDimensionTypes;
-import net.azisaba.vanilife.VanilifeLevelStems;
+import net.azisaba.vanilife.server.VanilifeBiomes;
+import net.azisaba.vanilife.server.VanilifeDimensionTypes;
+import net.azisaba.vanilife.server.VanilifeLevelStems;
+import net.azisaba.vanilife.server.VanilifeConfiguredFeatures;
+import net.azisaba.vanilife.server.VanilifePlacedFeatures;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
 import net.minecraft.CrashReport;
@@ -181,7 +182,6 @@ public class RegistryDataLoader {
         return load((loader, infoLookup) -> loader.loadFromNetwork(elements, resourceProvider, infoLookup), registryLookups, registryData);
     }
 
-    @Modified(Op.CHANGE)
     private static RegistryAccess.Frozen load(
         RegistryDataLoader.LoadingFunction loadingFunction,
         List<HolderLookup.RegistryLookup<?>> registryLookups,
@@ -193,19 +193,6 @@ public class RegistryDataLoader {
             .collect(Collectors.toUnmodifiableList());
         RegistryOps.RegistryInfoLookup registryInfoLookup = createContext(registryLookups, list);
         list.forEach(loader -> loadingFunction.apply((RegistryDataLoader.Loader<?>)loader, registryInfoLookup));
-        // Aetheria start - Register Aetheria objects in data-driven registries
-        list.forEach(loader -> {
-            final WritableRegistry<?> registry = loader.registry();
-
-            if (registry.key().equals(Registries.BIOME)) {
-                VanilifeBiomes.bootstrap((WritableRegistry<Biome>) registry, registryInfoLookup);
-            } else if (registry.key().equals(Registries.DIMENSION_TYPE)) {
-                VanilifeDimensionTypes.bootstrap((WritableRegistry<DimensionType>) registry, registryInfoLookup);
-            } else if (registry.key().equals(Registries.LEVEL_STEM)) {
-                VanilifeLevelStems.bootstrap((WritableRegistry<LevelStem>) registry, registryInfoLookup);
-            }
-        });
-        // Aetheria end - Register Aetheria objects in data-driven registries
         list.forEach(loader -> {
             Registry<?> registry = loader.registry();
 
@@ -398,7 +385,21 @@ public class RegistryDataLoader {
     }
 
     record Loader<T>(RegistryDataLoader.RegistryData<T> data, WritableRegistry<T> registry, Map<ResourceKey<?>, Exception> loadingErrors) {
+        @Modified(Op.CHANGE)
         public void loadFromResources(ResourceManager resourceManager, RegistryOps.RegistryInfoLookup registryInfoLookup) {
+            // Vanilife start - Register Vanilife objects in data-driven registries
+            if (registry.key().equals(Registries.BIOME)) {
+                VanilifeBiomes.bootstrap((WritableRegistry<Biome>) registry, registryInfoLookup);
+            } else if (registry.key().equals(Registries.CONFIGURED_FEATURE)) {
+                VanilifeConfiguredFeatures.bootstrap((WritableRegistry<ConfiguredFeature<?, ?>>) registry, registryInfoLookup);
+            } else if (registry.key().equals(Registries.PLACED_FEATURE)) {
+                VanilifePlacedFeatures.bootstrap((WritableRegistry<PlacedFeature>) registry, registryInfoLookup);
+            } else if (registry.key().equals(Registries.DIMENSION_TYPE)) {
+                VanilifeDimensionTypes.bootstrap((WritableRegistry<DimensionType>) registry, registryInfoLookup);
+            } else if (registry.key().equals(Registries.LEVEL_STEM)) {
+                VanilifeLevelStems.bootstrap((WritableRegistry<LevelStem>) registry, registryInfoLookup);
+            }
+            // Vanilife end - Register Vanilife objects in data-driven registries
             RegistryDataLoader.loadContentsFromManager(resourceManager, registryInfoLookup, this.registry, this.data.elementCodec, this.loadingErrors);
         }
 
diff --git a/net/minecraft/server/Bootstrap.java b/net/minecraft/server/Bootstrap.java
index ed675ee78be0c06c7b07ded7d3b9d87b2b4fdc15..efb850b013758fb550833926c5526312fa49d856 100644
--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -9,12 +9,14 @@ import java.util.TreeSet;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 import java.util.function.Supplier;
+import net.azisaba.vanilife.server.VanilifeFeatures;
 import net.minecraft.SharedConstants;
 import net.minecraft.SuppressForbidden;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.selector.options.EntitySelectorOptions;
 import net.minecraft.core.cauldron.CauldronInteraction;
 import net.minecraft.core.dispenser.DispenseItemBehavior;
+import net.minecraft.core.WritableRegistry;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.locale.Language;
 import net.minecraft.world.effect.MobEffect;
@@ -55,6 +57,7 @@ public class Bootstrap {
                     EntitySelectorOptions.bootStrap();
                     DispenseItemBehavior.bootStrap();
                     CauldronInteraction.bootStrap();
+                    VanilifeFeatures.bootstrap((WritableRegistry<net.minecraft.world.level.levelgen.feature.Feature<?>>) BuiltInRegistries.FEATURE);
                     // Paper start
                     BuiltInRegistries.bootStrap(() -> {
                         io.papermc.paper.world.worldgen.OptionallyFlatBedrockConditionSource.bootstrap(); // Paper - Flat bedrock generator settings
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index bdd3062d0e2cca647c749259c8d497d05f7fed86..578bad3c53aba748601a3b7686f7cef5355ba25c 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2464,7 +2464,7 @@ public class ServerGamePacketListenerImpl
             return;
         }
 
-        // Remove signed parts if the command was changed
+        // Remove signed conditionedParts if the command was changed
         if (!command.equals(packet.command())) {
             parseResults = this.parseCommand(command);
             map = Collections.emptyMap();
diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 08a1be57e1057a9e961eeb8f59cd12e98ed52bab..85f9767632b21d4194076b2205422d909791c56b 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -504,7 +504,7 @@ public class ServerExplosion implements Explosion {
                             // - Setting a velocity to a EnderDragonPart is ignored (and therefore not needed)
                             // - Damaging EnderDragonPart while forward the damage to EnderDragon
                             // - Damaging EnderDragon does nothing
-                            // - EnderDragon hitbox always covers the other parts and is therefore always present
+                            // - EnderDragon hitbox always covers the other conditionedParts and is therefore always present
                             if (entity instanceof EnderDragonPart) {
                                 continue;
                             }
diff --git a/net/minecraft/world/level/biome/BiomeManager.java b/net/minecraft/world/level/biome/BiomeManager.java
index 1ee7b32ae9dc579337520edfbf337ffaeaeebf00..ca31b8952ed1d7ecdd708f734e7aa6216f2d3740 100644
--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -1,8 +1,8 @@
 package net.minecraft.world.level.biome;
 
 import com.google.common.hash.Hashing;
-import net.azisaba.vanilife.world.AetheriaLayer;
-import net.azisaba.vanilife.world.AetheriaLayout;
+import net.azisaba.vanilife.server.world.resource.ResourceLayer;
+import net.azisaba.vanilife.server.world.resource.ResourceLayout;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.QuartPos;
@@ -31,7 +31,7 @@ public class BiomeManager {
     }
 
     // Aetheria start - Remap biome sampling Y by layer
-    public BiomeManager withLayeredSource(final AetheriaLayout layout, final AetheriaLayer.Type type) {
+    public BiomeManager withLayeredSource(final ResourceLayout layout, final ResourceLayer.Type type) {
         return this.withDifferentSource((quartX, quartY, quartZ) -> {
             final int layerY = QuartPos.toBlock(quartY);
             final int blockY = layout.toBlockY(type, layerY);
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index 492b1615d96133c5082e0f16015f114299f8082f..37045ec700915896c924d159075aa58fd2aceff1 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -21,6 +21,9 @@ import java.util.function.IntFunction;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.event.BlockDropLootEvent;
 import net.minecraft.SharedConstants;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -71,6 +74,10 @@ import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.block.CraftBlockStates;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -364,14 +371,26 @@ public class Block extends BlockBehaviour implements ItemLike {
     public void destroy(LevelAccessor level, BlockPos pos, BlockState state) {
     }
 
+    @Modified(Op.ADD_EVENT_CALL)
     public static List<ItemStack> getDrops(BlockState state, ServerLevel level, BlockPos pos, @Nullable BlockEntity blockEntity) {
         LootParams.Builder builder = new LootParams.Builder(level)
             .withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(pos))
             .withParameter(LootContextParams.TOOL, ItemStack.EMPTY)
             .withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockEntity);
-        return state.getDrops(builder);
+
+        final List<ItemStack> vanillaDrops = state.getDrops(builder);
+        final BlockDropLootEvent event = new BlockDropLootEvent(
+                CraftBlock.at(level, pos),
+                CraftBlockStates.getBlockState(level.getWorld(), pos, state, blockEntity),
+                null,
+                null,
+                vanillaDrops.stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList())
+        );
+        event.callEvent();
+        return event.getDrops().stream().map(CraftItemStack::asNMSCopy).toList();
     }
 
+    @Modified(Op.ADD_EVENT_CALL)
     public static List<ItemStack> getDrops(
         BlockState state, ServerLevel level, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack tool
     ) {
@@ -380,7 +399,17 @@ public class Block extends BlockBehaviour implements ItemLike {
             .withParameter(LootContextParams.TOOL, tool)
             .withOptionalParameter(LootContextParams.THIS_ENTITY, entity)
             .withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockEntity);
-        return state.getDrops(builder);
+
+        final List<ItemStack> vanillaDrops = state.getDrops(builder);
+        final BlockDropLootEvent event = new BlockDropLootEvent(
+                CraftBlock.at(level, pos),
+                CraftBlockStates.getBlockState(level.getWorld(), pos, state, blockEntity),
+                entity != null ? entity.getBukkitEntity() : null,
+                tool.asBukkitCopy(),
+                vanillaDrops.stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList())
+        );
+        event.callEvent();
+        return event.getDrops().stream().map(CraftItemStack::asNMSCopy).toList();
     }
 
     // Paper start - Add BlockBreakBlockEvent
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index c5953d5751e655e121e459aaab35e7719948e69e..79c41dbedc90eac928d27454ceb712472e83a8e6 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -15,6 +15,9 @@ import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.function.ToIntFunction;
 import java.util.stream.Stream;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
+import net.azisaba.vanilife.event.BlockRandomTickEvent;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
@@ -82,6 +85,8 @@ import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.util.RandomSourceWrapper;
 import org.jspecify.annotations.Nullable;
 
 public abstract class BlockBehaviour implements FeatureElement {
@@ -877,7 +882,13 @@ public abstract class BlockBehaviour implements FeatureElement {
             this.getBlock().tick(this.asState(), level, pos, random);
         }
 
+        @Modified(Op.ADD_EVENT_CALL)
         public void randomTick(ServerLevel level, BlockPos pos, RandomSource random) {
+            // Vanilife start - Add a call to net.azisaba.vanilife.event.BlockRandomTickEvent
+            final BlockRandomTickEvent event = new BlockRandomTickEvent(CraftBlock.at(level, pos), new RandomSourceWrapper.RandomWrapper(random));
+            if (!event.callEvent()) return;
+            // Vanilife end - Add a call to net.azisaba.vanilife.event.BlockRandomTickEvent
+
             this.getBlock().randomTick(this.asState(), level, pos, random);
         }
 
diff --git a/net/minecraft/world/level/chunk/ChunkGenerator.java b/net/minecraft/world/level/chunk/ChunkGenerator.java
index fd92282bb1cbe7051258f6d239e8dc0ce191baba..5d0d04f21e46d58c20badd2127fcd95c202d25c3 100644
--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -11,7 +11,7 @@ import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index d88c703e0eb22af1d815d712f4499305497c1180..8bf486b35f71a9a1642d28a03ee24cc201a664da 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -10,6 +10,8 @@ import java.util.EnumSet;
 import java.util.Set;
 import java.util.function.IntFunction;
 import java.util.function.Predicate;
+import net.azisaba.vanilife.annotations.Modified;
+import net.azisaba.vanilife.annotations.Op;
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
@@ -141,6 +143,7 @@ public class Heightmap {
         return x + z * 16;
     }
 
+    @Modified(Op.ADD_ENUM)
     public static enum Types implements StringRepresentable {
         WORLD_SURFACE_WG(0, "WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
         WORLD_SURFACE(1, "WORLD_SURFACE", Heightmap.Usage.CLIENT, Heightmap.NOT_AIR),
@@ -154,43 +157,43 @@ public class Heightmap {
             state -> (state.blocksMotion() || !state.getFluidState().isEmpty()) && !(state.getBlock() instanceof LeavesBlock)
         ),
 
-        // Aetheria start - Custom heightmap types by Aetheria layer
-        AETHERIA_OVERWORLD_WORLD_SURFACE_WG(6, "AETHERIA_OVERWORLD_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
-        AETHERIA_OVERWORLD_WORLD_SURFACE(7, "AETHERIA_OVERWORLD_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
-        AETHERIA_OVERWORLD_OCEAN_FLOOR_WG(8, "AETHERIA_OVERWORLD_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_OVERWORLD_OCEAN_FLOOR(9, "AETHERIA_OVERWORLD_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_OVERWORLD_MOTION_BLOCKING(10, "AETHERIA_OVERWORLD_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
-        AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES(
+        // Vanilife start - Custom heightmap types by resource layer
+        RESOURCE_OVERWORLD_WORLD_SURFACE_WG(6, "RESOURCE_OVERWORLD_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
+        RESOURCE_OVERWORLD_WORLD_SURFACE(7, "RESOURCE_OVERWORLD_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
+        RESOURCE_OVERWORLD_OCEAN_FLOOR_WG(8, "RESOURCE_OVERWORLD_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_OVERWORLD_OCEAN_FLOOR(9, "RESOURCE_OVERWORLD_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_OVERWORLD_MOTION_BLOCKING(10, "RESOURCE_OVERWORLD_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
+        RESOURCE_OVERWORLD_MOTION_BLOCKING_NO_LEAVES(
                 11,
-                "AETHERIA_OVERWORLD_MOTION_BLOCKING_NO_LEAVES",
+                "RESOURCE_OVERWORLD_MOTION_BLOCKING_NO_LEAVES",
                 Heightmap.Usage.LIVE_WORLD,
                 state -> (state.blocksMotion() || !state.getFluidState().isEmpty()) && !(state.getBlock() instanceof LeavesBlock)
         ),
 
-        AETHERIA_NETHER_WORLD_SURFACE_WG(12, "AETHERIA_NETHER_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
-        AETHERIA_NETHER_WORLD_SURFACE(13, "AETHERIA_NETHER_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
-        AETHERIA_NETHER_OCEAN_FLOOR_WG(14, "AETHERIA_NETHER_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_NETHER_OCEAN_FLOOR(15, "AETHERIA_NETHER_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_NETHER_MOTION_BLOCKING(16, "AETHERIA_NETHER_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
-        AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES(
+        RESOURCE_NETHER_WORLD_SURFACE_WG(12, "RESOURCE_NETHER_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
+        RESOURCE_NETHER_WORLD_SURFACE(13, "RESOURCE_NETHER_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
+        RESOURCE_NETHER_OCEAN_FLOOR_WG(14, "RESOURCE_NETHER_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_NETHER_OCEAN_FLOOR(15, "RESOURCE_NETHER_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_NETHER_MOTION_BLOCKING(16, "RESOURCE_NETHER_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
+        RESOURCE_NETHER_MOTION_BLOCKING_NO_LEAVES(
                 17,
-                "AETHERIA_NETHER_MOTION_BLOCKING_NO_LEAVES",
+                "RESOURCE_NETHER_MOTION_BLOCKING_NO_LEAVES",
                 Heightmap.Usage.LIVE_WORLD,
                 state -> (state.blocksMotion() || !state.getFluidState().isEmpty()) && !(state.getBlock() instanceof LeavesBlock)
         ),
 
-        AETHERIA_END_WORLD_SURFACE_WG(18, "AETHERIA_END_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
-        AETHERIA_END_WORLD_SURFACE(19, "AETHERIA_END_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
-        AETHERIA_END_OCEAN_FLOOR_WG(20, "AETHERIA_END_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_END_OCEAN_FLOOR(21, "AETHERIA_END_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
-        AETHERIA_END_MOTION_BLOCKING(22, "AETHERIA_END_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
-        AETHERIA_END_MOTION_BLOCKING_NO_LEAVES(
+        RESOURCE_END_WORLD_SURFACE_WG(18, "RESOURCE_END_WORLD_SURFACE_WG", Heightmap.Usage.WORLDGEN, Heightmap.NOT_AIR),
+        RESOURCE_END_WORLD_SURFACE(19, "RESOURCE_END_WORLD_SURFACE", Heightmap.Usage.LIVE_WORLD, Heightmap.NOT_AIR),
+        RESOURCE_END_OCEAN_FLOOR_WG(20, "RESOURCE_END_OCEAN_FLOOR_WG", Heightmap.Usage.WORLDGEN, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_END_OCEAN_FLOOR(21, "RESOURCE_END_OCEAN_FLOOR", Heightmap.Usage.LIVE_WORLD, Heightmap.MATERIAL_MOTION_BLOCKING),
+        RESOURCE_END_MOTION_BLOCKING(22, "RESOURCE_END_MOTION_BLOCKING", Heightmap.Usage.LIVE_WORLD, state -> state.blocksMotion() || !state.getFluidState().isEmpty()),
+        RESOURCE_END_MOTION_BLOCKING_NO_LEAVES(
                 23,
-                "AETHERIA_END_MOTION_BLOCKING_NO_LEAVES",
+                "RESOURCE_END_MOTION_BLOCKING_NO_LEAVES",
                 Heightmap.Usage.LIVE_WORLD,
                 state -> (state.blocksMotion() || !state.getFluidState().isEmpty()) && !(state.getBlock() instanceof LeavesBlock)
         );
-        // Aetheria end - Custom heightmap types by Aetheria layer
+        // Vanilife end - Custom heightmap types by resource layer
 
         public static final Codec<Heightmap.Types> CODEC = StringRepresentable.fromEnum(Heightmap.Types::values);
         private static final IntFunction<Heightmap.Types> BY_ID = ByIdMap.continuous(types -> types.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
diff --git a/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java b/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
index f12520f990d28c28e8ce8d4c3c182da266f159b6..4e289e473205b2df162314aad2ef45b1d35a14eb 100644
--- a/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
@@ -3,7 +3,7 @@ package net.minecraft.world.level.levelgen.feature;
 import com.mojang.serialization.Codec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
diff --git a/net/minecraft/world/level/levelgen/feature/Feature.java b/net/minecraft/world/level/levelgen/feature/Feature.java
index cf41eaeec0eea4e320dfc1d820772ef63217bebb..c2b5c5cb381954e3cabe5d48679da5494d9964f3 100644
--- a/net/minecraft/world/level/levelgen/feature/Feature.java
+++ b/net/minecraft/world/level/levelgen/feature/Feature.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Registry;
diff --git a/net/minecraft/world/level/levelgen/feature/FeaturePlaceContext.java b/net/minecraft/world/level/levelgen/feature/FeaturePlaceContext.java
index a544df25b1167084a8f898463da68ffb3377f05e..6c893e94d70e5549e0ba9c93f40d53371820370d 100644
--- a/net/minecraft/world/level/levelgen/feature/FeaturePlaceContext.java
+++ b/net/minecraft/world/level/levelgen/feature/FeaturePlaceContext.java
@@ -2,7 +2,7 @@ package net.minecraft.world.level.levelgen.feature;
 
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.WorldGenLevel;
diff --git a/net/minecraft/world/level/levelgen/feature/LargeDripstoneFeature.java b/net/minecraft/world/level/levelgen/feature/LargeDripstoneFeature.java
index ecb573ca0521d86bfda9ea793c391c01ae71f5e6..05588383b6dff2ce95ba60fac27ca6b35a8988c8 100644
--- a/net/minecraft/world/level/levelgen/feature/LargeDripstoneFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/LargeDripstoneFeature.java
@@ -4,7 +4,7 @@ import java.util.Optional;
 import com.mojang.serialization.Codec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.SharedConstants;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
diff --git a/net/minecraft/world/level/levelgen/placement/BiomeFilter.java b/net/minecraft/world/level/levelgen/placement/BiomeFilter.java
index 4e07c03af2f117634477db3de54fa5a046d04bac..b06a8a634740b39e39079a1324d9b65ecadff3e5 100644
--- a/net/minecraft/world/level/levelgen/placement/BiomeFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/BiomeFilter.java
@@ -3,7 +3,7 @@ package net.minecraft.world.level.levelgen.placement;
 import com.mojang.serialization.MapCodec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.util.RandomSource;
diff --git a/net/minecraft/world/level/levelgen/placement/BlockPredicateFilter.java b/net/minecraft/world/level/levelgen/placement/BlockPredicateFilter.java
index 7aed7a8c3a67ab03b93469e0e7b875c1df1543b4..8ed12dc2b9991088f48513128c9825f25e46704a 100644
--- a/net/minecraft/world/level/levelgen/placement/BlockPredicateFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/BlockPredicateFilter.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.levelgen.blockpredicates.BlockPredicate;
diff --git a/net/minecraft/world/level/levelgen/placement/CountOnEveryLayerPlacement.java b/net/minecraft/world/level/levelgen/placement/CountOnEveryLayerPlacement.java
index 4c5c8659a77606acb412bd9fc26aae9b8ec9e175..611b17270c3ffe90de7f05e80bd7875f4859a054 100644
--- a/net/minecraft/world/level/levelgen/placement/CountOnEveryLayerPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/CountOnEveryLayerPlacement.java
@@ -3,7 +3,7 @@ package net.minecraft.world.level.levelgen.placement;
 import com.mojang.serialization.MapCodec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.ConstantInt;
diff --git a/net/minecraft/world/level/levelgen/placement/EnvironmentScanPlacement.java b/net/minecraft/world/level/levelgen/placement/EnvironmentScanPlacement.java
index cd9e90a0b3da965e691c594b5297451c320a8b14..cd4fc90154e28e5054730943b92d4faf0a21aa64 100644
--- a/net/minecraft/world/level/levelgen/placement/EnvironmentScanPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/EnvironmentScanPlacement.java
@@ -5,7 +5,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.util.RandomSource;
diff --git a/net/minecraft/world/level/levelgen/placement/FixedPlacement.java b/net/minecraft/world/level/levelgen/placement/FixedPlacement.java
index f24f6b81f9df5aa85534ad6da2794dda072fdfdf..fd4a2073023d678ed58eb0a5878087e97365c3c5 100644
--- a/net/minecraft/world/level/levelgen/placement/FixedPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/FixedPlacement.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.util.RandomSource;
diff --git a/net/minecraft/world/level/levelgen/placement/HeightRangePlacement.java b/net/minecraft/world/level/levelgen/placement/HeightRangePlacement.java
index 30925091a3b1dfeb47992ee93c95b5e24ca566a9..1c4ed9be9c0a20b470282aef45ae9fc804e09f1b 100644
--- a/net/minecraft/world/level/levelgen/placement/HeightRangePlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/HeightRangePlacement.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.levelgen.VerticalAnchor;
diff --git a/net/minecraft/world/level/levelgen/placement/HeightmapPlacement.java b/net/minecraft/world/level/levelgen/placement/HeightmapPlacement.java
index 2cdafec2c0cfa4c700e3cabd664442d5d49efcba..3043dc001b447b5589d86041d560871aab82faa3 100644
--- a/net/minecraft/world/level/levelgen/placement/HeightmapPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/HeightmapPlacement.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.levelgen.Heightmap;
diff --git a/net/minecraft/world/level/levelgen/placement/InSquarePlacement.java b/net/minecraft/world/level/levelgen/placement/InSquarePlacement.java
index 1d8e98c381a0d805db68d29beb5bb7570e8ad317..eaaeeb25798f3d469033ed1b46df5e281a0bc07c 100644
--- a/net/minecraft/world/level/levelgen/placement/InSquarePlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/InSquarePlacement.java
@@ -3,7 +3,7 @@ package net.minecraft.world.level.levelgen.placement;
 import com.mojang.serialization.MapCodec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 
diff --git a/net/minecraft/world/level/levelgen/placement/PlacedFeature.java b/net/minecraft/world/level/levelgen/placement/PlacedFeature.java
index 5ec2bea9751d4f4a1e1c968555dc7901603d1f8f..11f26a2e241766746b587f63c5285b16e5612bf6 100644
--- a/net/minecraft/world/level/levelgen/placement/PlacedFeature.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacedFeature.java
@@ -8,7 +8,7 @@ import java.util.stream.Stream;
 
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.SharedConstants;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
diff --git a/net/minecraft/world/level/levelgen/placement/PlacementFilter.java b/net/minecraft/world/level/levelgen/placement/PlacementFilter.java
index 44a2d1086f0212d6c066a754e8c6bbc00f4ce3c8..c8bb615bf1383a1d5c6e28365203a8d8eb8dfd2b 100644
--- a/net/minecraft/world/level/levelgen/placement/PlacementFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacementFilter.java
@@ -2,7 +2,7 @@ package net.minecraft.world.level.levelgen.placement;
 
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 
diff --git a/net/minecraft/world/level/levelgen/placement/PlacementModifier.java b/net/minecraft/world/level/levelgen/placement/PlacementModifier.java
index 2fe1113e8ae2a85d151953ac60084aeadecb5359..216ec569479b9feab9c760cad99c8da607cc2666 100644
--- a/net/minecraft/world/level/levelgen/placement/PlacementModifier.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacementModifier.java
@@ -5,7 +5,7 @@ import java.util.stream.Stream;
 
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.util.RandomSource;
diff --git a/net/minecraft/world/level/levelgen/placement/RandomOffsetPlacement.java b/net/minecraft/world/level/levelgen/placement/RandomOffsetPlacement.java
index 9df3f4ace9fbd54055bdd14b8d886987f910e9c3..ec1f3659da013d3170db04eda3d26fc5c1e81e29 100644
--- a/net/minecraft/world/level/levelgen/placement/RandomOffsetPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/RandomOffsetPlacement.java
@@ -4,7 +4,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.ConstantInt;
diff --git a/net/minecraft/world/level/levelgen/placement/RarityFilter.java b/net/minecraft/world/level/levelgen/placement/RarityFilter.java
index d09e0bfea96e3d9093361cbdebe3064d3e83ee92..226a15ba7051b841702460f7dd73989f37d1f1bd 100644
--- a/net/minecraft/world/level/levelgen/placement/RarityFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/RarityFilter.java
@@ -3,7 +3,7 @@ package net.minecraft.world.level.levelgen.placement;
 import com.mojang.serialization.MapCodec;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
diff --git a/net/minecraft/world/level/levelgen/placement/RepeatingPlacement.java b/net/minecraft/world/level/levelgen/placement/RepeatingPlacement.java
index 1ab1edfe58426f806496be816ffdfe0f98404bd3..9b3c9bc3cef82171602bda394dfbab68d64a8e3f 100644
--- a/net/minecraft/world/level/levelgen/placement/RepeatingPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/RepeatingPlacement.java
@@ -2,7 +2,7 @@ package net.minecraft.world.level.levelgen.placement;
 
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 
diff --git a/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java b/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java
index 5d00477ad63ba7cc843674c3e423f98d70e3ea34..92b4f8c907f933b78f75ae360515556d3b20bade 100644
--- a/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java
@@ -5,7 +5,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.levelgen.Heightmap;
diff --git a/net/minecraft/world/level/levelgen/placement/SurfaceWaterDepthFilter.java b/net/minecraft/world/level/levelgen/placement/SurfaceWaterDepthFilter.java
index 51e0290e9d1b0e81174260d9b5470ce5d2e2db54..85b32eafcd45dc5c9e881c328ca8670e21e8926a 100644
--- a/net/minecraft/world/level/levelgen/placement/SurfaceWaterDepthFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/SurfaceWaterDepthFilter.java
@@ -5,7 +5,7 @@ import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.azisaba.vanilife.annotations.Modified;
 import net.azisaba.vanilife.annotations.Op;
-import net.azisaba.vanilife.world.height.HeightContext;
+import net.azisaba.vanilife.server.world.height.HeightContext;
 import net.minecraft.core.BlockPos;
 import net.minecraft.util.RandomSource;
 
diff --git a/net/minecraft/world/phys/shapes/Shapes.java b/net/minecraft/world/phys/shapes/Shapes.java
index fc2b393d24fef8ce3a74d38d0937bfe5604eae87..e516e5f19835556887069369c38e18d849732685 100644
--- a/net/minecraft/world/phys/shapes/Shapes.java
+++ b/net/minecraft/world/phys/shapes/Shapes.java
@@ -64,8 +64,8 @@ public final class Shapes {
     };
 
     private static double[] generateCubeParts(final int parts) {
-        // note: parts is a power of two, so we do not need to worry about loss of precision here
-        // note: parts is from [2^0, 2^3]
+        // note: conditionedParts is a power of two, so we do not need to worry about loss of precision here
+        // note: conditionedParts is from [2^0, 2^3]
         final double inc = 1.0 / (double)parts;
 
         final double[] ret = new double[parts + 1];
