From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Fri, 20 Dec 2024 12:41:13 -0800
Subject: [PATCH] Use Folia logo


diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index 023cca49c6dd0f2961cf30f02c7cd74601f73f8b..79518cbca7760f710130509d93d27fb017d67159 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -58,7 +58,7 @@ public class MinecraftServerGui extends JComponent {
         // Paper start - Improve ServerGUI
         frame.setName("Minecraft server");
         try {
-            frame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
+            frame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("folia-logo.png"))));
         } catch (java.io.IOException ignore) {
         }
         // Paper end - Improve ServerGUI
