--- a/paper-checkstyle/build.gradle.kts
+++ b/paper-checkstyle/build.gradle.kts
@@ -5,6 +_,19 @@
 
 val testData = sourceSets.create("testData")
 
+tasks.named<io.papermc.paperweight.checkstyle.tasks.MergeCheckstyleConfigs>("mergeCheckstyleConfigs") {
+    overrideConfigFile.set(layout.projectDirectory.file("../paper-checkstyle/.checkstyle/checkstyle.xml"))
+}
+
+sourceSets {
+    main {
+        java.srcDir(file("../paper-checkstyle/src/main/java"))
+        resources.srcDir(file("../paper-checkstyle/src/main/resources"))
+    }
+    test {
+        java.srcDir(file("../paper-checkstyle/src/test/java"))
+        resources.srcDir(file("../paper-checkstyle/src/test/resources"))
+    }
+}
+testData.java.srcDir(file("../paper-checkstyle/src/testData/java"))
+testData.resources.srcDir(file("../paper-checkstyle/src/testData/resources"))
+
 dependencies {
     implementation("com.puppycrawl.tools:checkstyle:13.8.0")
     implementation("org.jspecify:jspecify:1.0.0")
@@ -16,11 +_,12 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
     testData.implementationConfigurationName("org.jspecify:jspecify:1.0.0")
 
-    checkstyle(project(":paper-checkstyle"))
+    checkstyle(project(":folia-checkstyle"))
 }
 
 tasks {
     test {
+        workingDir = file("../paper-checkstyle")
         useJUnitPlatform()
     }
 }
