From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MiniDigger <admin@minidigger.me>
Date: Sat, 12 Jun 2021 16:40:34 +0200
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index f0256fe4aea16c9e0d6f5eee360eeec36cfdeebd..8a13deacc6d70ae1555f1cc8dedd1b7cb50bf291 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,7 +20,7 @@ repositories {
 }
 
 dependencies {
-    implementation(project(":Paper-API"))
+    implementation(project(":ForkTest-API"))
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.12.1")
     implementation("net.minecrell:terminalconsoleappender:1.2.0")
@@ -98,7 +98,7 @@ tasks.test {
 
 fun TaskContainer.registerRunTask(name: String, block: JavaExec.() -> Unit): TaskProvider<JavaExec> =
     register<JavaExec>(name) {
-        group = "Paper"
+        group = "paperweight"
         workingDir = rootProject.layout.projectDirectory.dir(
             providers.gradleProperty("runWorkDir").forUseAtConfigurationTime().orElse("run")
         ).get().asFile
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index cf92c3275869e4a0209fd4b07ad723e11717dc86..9846c65bbd77e5e30dcb93a90cdd961e58244ba2 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -459,7 +459,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 DedicatedServerProperties dedicatedserverproperties = ((DedicatedServer) this).getProperties();
 
                 worldsettings = new LevelSettings(dedicatedserverproperties.levelName, dedicatedserverproperties.gamemode, dedicatedserverproperties.hardcore, dedicatedserverproperties.difficulty, false, new GameRules(), this.datapackconfiguration);
-                generatorsettings = this.options.has("bonusChest") ? dedicatedserverproperties.worldGenSettings.withBonusChest() : dedicatedserverproperties.worldGenSettings;
+                generatorsettings = this.options.has("bonusChest") ? dedicatedserverproperties.getWorldGenSettings(iregistrycustom_dimension).withBonusChest() : dedicatedserverproperties.getWorldGenSettings(iregistrycustom_dimension);
             }
 
             overworldData = new PrimaryLevelData(worldsettings, generatorsettings, Lifecycle.stable());
@@ -548,7 +548,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                     DedicatedServerProperties dedicatedserverproperties = ((DedicatedServer) this).getProperties();
 
                     worldsettings = new LevelSettings(dedicatedserverproperties.levelName, dedicatedserverproperties.gamemode, dedicatedserverproperties.hardcore, dedicatedserverproperties.difficulty, false, new GameRules(), this.datapackconfiguration);
-                    generatorsettings = this.options.has("bonusChest") ? dedicatedserverproperties.worldGenSettings.withBonusChest() : dedicatedserverproperties.worldGenSettings;
+                    generatorsettings = this.options.has("bonusChest") ? dedicatedserverproperties.getWorldGenSettings(iregistrycustom_dimension).withBonusChest() : dedicatedserverproperties.getWorldGenSettings(iregistrycustom_dimension);
                 }
 
                 worlddata = new PrimaryLevelData(worldsettings, generatorsettings, Lifecycle.stable());
@@ -1511,7 +1511,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "Paper"; //Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        return "ForkTest"; //Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
     }
 
     public SystemReport fillSystemReport(SystemReport systemreport) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3452fc8ccddd5e9c8c12d98f55473fb2e322859a..58098f30f94e27112e23b4873bbd112a3dc8b546 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -230,7 +230,7 @@ import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
-    private final String serverName = "Paper"; // Paper
+    private final String serverName = "ForkTest"; // Paper
     private final String serverVersion;
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
