From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Sun, 12 Mar 2023 00:51:29 -0800
Subject: [PATCH] fixup! Increase parallelism for neighbour writing chunk
 statuses


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/queue/RadiusAwarePrioritisedExecutor.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/queue/RadiusAwarePrioritisedExecutor.java
index 1c8e05d5dde630fdee01900823d26c293f375abc..3272f73013ea7d4efdd0ae2903925cc543be7075 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/queue/RadiusAwarePrioritisedExecutor.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/queue/RadiusAwarePrioritisedExecutor.java
@@ -428,7 +428,7 @@ public class RadiusAwarePrioritisedExecutor {
         }
 
         private boolean isFiniteRadius() {
-            return this.radius > 0;
+            return this.radius >= 0;
         }
 
         private PrioritisedExecutor.PrioritisedTask pushTask(final PrioritisedExecutor executor) {
